p2p/protocol/identify/obsaddr.go

```diff
@@ -41,60 +46,85 @@ func (oa *ObservedAddr) TryActivate(ttl time.Duration) bool {
 type ObservedAddrSet struct {
 	sync.Mutex // guards whole datastruct.
 
-	addrs map[string]*ObservedAddr
+	// local(internal) address -> list of observed(external) addresses
+	addrs map[string][]*ObservedAddr
 	ttl   time.Duration
 }
 
-func (oas *ObservedAddrSet) Addrs() []ma.Multiaddr {
+// Addrs return all activated observed addresses
+func (oas *ObservedAddrSet) Addrs() (addrs []ma.Multiaddr) {
 	oas.Lock()
 	defer oas.Unlock()
 
 	// for zero-value.
-	if oas.addrs == nil {
+	if len(oas.addrs) == 0 {
 		return nil
 	}
 
 	now := time.Now()
-	addrs := make([]ma.Multiaddr, 0, len(oas.addrs))
-	for s, a := range oas.addrs {
-		// remove timed out addresses.
-		if now.Sub(a.LastSeen) > oas.ttl {
-			delete(oas.addrs, s)
-			continue
-		}
-
-		if a.Activated || a.TryActivate(oas.ttl) {
-			addrs = append(addrs, a.Addr)
+	filteredAddrMap := make(map[string][]*ObservedAddr)
+	for local, observedAddrs := range oas.addrs {
+		filteredAddrs := make([]*ObservedAddr, 0, len(observedAddrs))
+		for _, a := range observedAddrs {
+			// leave only alive observed addresses
+			if now.Sub(a.LastSeen) <= oas.ttl {
+				filteredAddrs = append(filteredAddrs, a)
+				if a.activated(oas.ttl) {
+					addrs = append(addrs, a.Addr)
+				}
+			}
 		}
+		filteredAddrMap[local] = filteredAddrs
 	}
+	oas.addrs = filteredAddrMap
 	return addrs
 }
 
-func (oas *ObservedAddrSet) Add(addr ma.Multiaddr, observer ma.Multiaddr) {
+func (oas *ObservedAddrSet) Add(observed, local, observer ma.Multiaddr,
+	direction net.Direction) {
+
 	oas.Lock()
 	defer oas.Unlock()
 
 	// for zero-value.
 	if oas.addrs == nil {
-		oas.addrs = make(map[string]*ObservedAddr)
+		oas.addrs = make(map[string][]*ObservedAddr)
 		oas.ttl = pstore.OwnObservedAddrTTL
 	}
 
-	s := addr.String()
-	oa, found := oas.addrs[s]
+	now := time.Now()
+	observerString := observerGroup(observer)
+	localString := local.String()
+	observedAddr := &ObservedAddr{
+		Addr: observed,
+		SeenBy: map[string]observation{
+			observerString: {
+				seenTime:      now,
+				connDirection: direction,
+			},
+		},
+		LastSeen: now,
+	}
```

Comment:
This logic feels a bit out-of-order; we're constructing something we may not use. How about:

```
// construct an `observation` (not the full `observedAddr`).

observedAddrs := oas.addrs[localString] // no need for "found"

// Will be a no-op if observedAddrs is nil
for i, previousObserved := range observedAddrs {
    // ...
}

// Now construct the full `ObservedAddr` and insert it.
```

I haven't tried reordering this so it may not be any cleaner but I *think* it'll be a bit easier to follow.